Schema registry 4/N

mgodave opened the pull request:
See #1137 for reference

merlimat commented on pulsar-broker/src/main/java/org/apache/pulsar/broker/admin/SchemasResource.java:
```diff
@@ -0,0 +1,273 @@
+import org.apache.pulsar.common.schema.SchemaVersion;
+
+@Path("/schemas")
+public class SchemasResource extends AdminResource {
```
This handler should go on the v2 admin API

merlimat commented on pulsar-broker/src/main/java/org/apache/pulsar/broker/admin/SchemasResource.java:
```diff
@@ -0,0 +1,273 @@
+        this(Clock.systemUTC());
+    }
+
+    @GET @Path("/{property}/{cluster}/{namespace}/{topic}/schema")
```
For existing REST API handlers, we have split the endpoints in 2: 
 * One with the old topic name 
 * One with v2 topic name (no cluster in the name).

I think that for this new feature we could just concentrate on v2 API.

merlimat commented on pulsar-broker/src/main/java/org/apache/pulsar/broker/admin/SchemasResource.java:
```diff
@@ -0,0 +1,273 @@
+    }
+
+    private String buildSchemaId(String property, String cluster, String namespace, String topic) {
+        return property + "_" + cluster + "_" + namespace + "_" + topic;
```
I think we already added a method to get the schema in `TopicName`, right?

merlimat commented on pulsar-broker/src/main/java/org/apache/pulsar/broker/admin/SchemasResource.java:
```diff
@@ -0,0 +1,273 @@
+        }
+    }
+
+    static class GetSchemaResponse {
```
If this classes will have to be used in the `PulsarAdmin` Java client, they should be added in `pulsar-common` module.

mgodave commented:
Hold on, I've made a lot of these changes but I think my last two branches
have misplaced commits. I might just merge them into one last pr.

On Wed, Mar 14, 2018 at 11:34 AM, Matteo Merli <notifications@github.com>
wrote:

> *@merlimat* commented on this pull request.
> ------------------------------
>
> In pulsar-broker/src/main/java/org/apache/pulsar/broker/
> admin/SchemasResource.java
> <https://github.com/apache/incubator-pulsar/pull/1381#discussion_r174545638>
> :
>
> > +import javax.ws.rs.Path;
> +import javax.ws.rs.PathParam;
> +import javax.ws.rs.Produces;
> +import javax.ws.rs.container.AsyncResponse;
> +import javax.ws.rs.container.Suspended;
> +import javax.ws.rs.core.MediaType;
> +import javax.ws.rs.core.Response;
> +import org.apache.pulsar.broker.service.Topic;
> +import org.apache.pulsar.broker.web.RestException;
> +import org.apache.pulsar.common.naming.TopicName;
> +import org.apache.pulsar.common.schema.SchemaData;
> +import org.apache.pulsar.common.schema.SchemaType;
> +import org.apache.pulsar.common.schema.SchemaVersion;
> +
> +@Path("/schemas")
> +public class SchemasResource extends AdminResource {
>
> This handler should go on the v2 admin API
> ------------------------------
>
> In pulsar-broker/src/main/java/org/apache/pulsar/broker/
> admin/SchemasResource.java
> <https://github.com/apache/incubator-pulsar/pull/1381#discussion_r174546254>
> :
>
> > +@Path("/schemas")
> +public class SchemasResource extends AdminResource {
> +
> +    private final Clock clock;
> +
> +    @VisibleForTesting
> +    SchemasResource(Clock clock) {
> +        super();
> +        this.clock = clock;
> +    }
> +
> +    public SchemasResource() {
> +        this(Clock.systemUTC());
> +    }
> +
> +    @GET @Path("/{property}/{cluster}/{namespace}/{topic}/schema")
>
> For existing REST API handlers, we have split the endpoints in 2:
>
>    - One with the old topic name
>    - One with v2 topic name (no cluster in the name).
>
> I think that for this new feature we could just concentrate on v2 API.
> ------------------------------
>
> In pulsar-broker/src/main/java/org/apache/pulsar/broker/
> admin/SchemasResource.java
> <https://github.com/apache/incubator-pulsar/pull/1381#discussion_r174546539>
> :
>
> > +                .isDeleted(false)
> +                .timestamp(clock.millis())
> +                .type(SchemaType.valueOf(payload.type))
> +                .user(clientAppId())
> +                .build()
> +        ).thenAccept(version ->
> +            response.resume(
> +                Response.accepted().entity(
> +                    new PostSchemaResponse(version)
> +                ).build()
> +            )
> +        );
> +    }
> +
> +    private String buildSchemaId(String property, String cluster, String namespace, String topic) {
> +        return property + "_" + cluster + "_" + namespace + "_" + topic;
>
> I think we already added a method to get the schema in TopicName, right?
> ------------------------------
>
> In pulsar-broker/src/main/java/org/apache/pulsar/broker/
> admin/SchemasResource.java
> <https://github.com/apache/incubator-pulsar/pull/1381#discussion_r174546852>
> :
>
> > +            } else {
> +                throw e;
> +            }
> +        }
> +    }
> +
> +    private void validateDestinationExists(TopicName dn) {
> +        try {
> +            Topic topic = pulsar().getBrokerService().getTopicReference(dn.toString());
> +            checkNotNull(topic);
> +        } catch (Exception e) {
> +            throw new RestException(Response.Status.NOT_FOUND, "Topic not found");
> +        }
> +    }
> +
> +    static class GetSchemaResponse {
>
> If this classes will have to be used in the PulsarAdmin Java client, they
> should be added in pulsar-common module.
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/apache/incubator-pulsar/pull/1381#pullrequestreview-103922460>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AAh9FtDtudZWOI4D9-5tpC5F8snJJps1ks5teVTBgaJpZM4SqqD8>
> .
>

sijie commented:
@mgodave which one I should review? this one or #1137 ?

mgodave commented:
this one, #1137 is just the tracking branch

On Mon, Mar 19, 2018 at 2:24 PM, Sijie Guo <notifications@github.com> wrote:

> @mgodave <https://github.com/mgodave> which one I should review? this one
> or #1137 <https://github.com/apache/incubator-pulsar/pull/1137> ?
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/apache/incubator-pulsar/pull/1381#issuecomment-374360569>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AAh9FnG6SKIVsLTF8rDaieRXQotFaDuDks5tgBPogaJpZM4SqqD8>
> .
>

merlimat commented on pulsar-client/src/main/java/org/apache/pulsar/client/api/Schema.java:
```diff
@@ -18,10 +18,14 @@
     byte[] encode(T message);
     T decode(byte[] bytes);
 
+    SchemaInfo getSchemaInfo();
```
If `SchemaInfo` becomes part of API, we should move it in the same `..client.api` package

mgodave replied:
It's used in multiple places. Are you suggesting I keep it in common but add it to the client package?

merlimat commented on pulsar-client/src/main/java/org/apache/pulsar/client/impl/ClientCnx.java:
```diff
@@ -77,7 +79,7 @@
     private State state;
 
-    private final ConcurrentLongHashMap<CompletableFuture<Pair<String, Long>>> pendingRequests = new ConcurrentLongHashMap<>(
+    private final ConcurrentLongHashMap<CompletableFuture<Triple<String, Long, byte[]>>> pendingRequests = new ConcurrentLongHashMap<>(
```
Maybe we could switch to a named class instead of the `Triple` so that it's easier to follow the meaning of each of them.

mgodave replied:
Yeah, I had a similar thought but took the easy way out :-( I'll add a named class.

merlimat commented on pulsar-common/src/main/proto/PulsarApi.proto:
```diff
@@ -398,7 +397,6 @@ message CommandRedeliverUnacknowledgedMessages {
 
 message CommandSuccess {
 	required uint64 request_id = 1;
-	optional Schema schema = 2;
```
Wasn't this needed to get the consumer to be able to auto-discover the schema at the subscribe time?

merlimat commented on pulsar-common/src/main/proto/PulsarApi.proto:
```diff
@@ -31,7 +31,6 @@ message Schema {
 	}
 
     required string name = 1;
-    required bytes version = 2;
```
Shouldn't we need to know the version in some cases? 

I'd imagine that a producer deriving the schema from a class type, won't know the version. But a consumer receiving a schema, might need the version?

merlimat commented:
@mgodave  there's a file that's missing the license header: `ProducerResponse.java`

merlimat commented:
@mgodave please merge with master again and it should go through CI this time.

merlimat commented:
@mgodave there is a conflict with a change that went into master again, sorry. 

I had changed `getTopicReference()` to return `Optional<Topic>`

sijie commented:
retest this please
